### ISSUE TITLE
Cassandra on Kubernetes - Seed Provider / Snitch

### ISSUE THREAD

**Issue by chrislovecnm** [kubernetes/kubernetes#24290]
Custom Cassandra makes every Cassandra node in a ring a seed provider.  The provider calls the K8s API and gets every Cassandra endpoint in the pod.  This data should be a small percentage of the nodes, and is static data.

For example for node layout
node1 - 10.10.0.1
node2 - 10.10.0.2
node3 - 10.10.0.3
node4 - 10.10.0.4

A configurable amount of nodes should be returned.  If that number of nodes is set to two, node1 and node2 ip addresses should be returned.  If node2 is replaced with a new IP address, that new ip address should be replaced.  Further investigation needs to occur.

**Comment by bgrant0607** [comment]
cc @kubernetes/examples

**Comment by chrislovecnm** [comment]
@bgrant0607 so good news ... This is a non-issue 👍   We do need unit testing, logging, more documentation, and a possible refactor ...

But the design is stable!!

**Comment by chrislovecnm** [comment]
@bgrant0607 please close, as I will open a ticket for all the other stuff!!

**Comment by bprashanth** [comment]
>A configurable amount of nodes should be returned. If that number of nodes is set to two, node1 and node2 ip addresses should be returned. If node2 is replaced with a new IP address, that new ip address should be replaced. Further investigation needs to occur.

This can happen, assuming it's a non issue somehow:

You get: 
node1 - 10.10.0.1
node2 - 10.10.0.2

Node2 dies and comes up with 10.10.0.6
Node7 form another c* cluster comes up with 10.10.0.2

you still continue to talk to Node7 thinking it's node2

Some time laters you see a new list:
node1 - 10.10.0.1
node2 - 10.10.0.7

**Comment by chrislovecnm** [comment]
So talking to a node is gossip 😄   not seeds ...  So when a node starts it needs a valid list of seeds.  If gossip starts to FREAK out, it will fall back to seeds.  

The seeds changing is ok.  As long as it is a valid ip address.  We need to look at the Snitch / Gossip soon.

**Comment by chrislovecnm** [comment]
Let me eat my words a bit. We'd need to cap it at 20 ... Say we have 1k c* nodes ... How much load will that put on the API server? I know of rings that have 3k c* nodes.

Cal we just ask the API for 3 servers?

**Comment by chrislovecnm** [comment]
@bgrant0607 @bprashanth so this is the challenge.  We don't want this code to make an http call to get 1000's of ip addresses.  Who is an api guru that can give some advice?

Much thanks!!

**Comment by bgrant0607** [comment]
@chrislovecnm If you could get the number of replicas and the hostname prefix, it seems like you could compute a fixed set of nodes and then just poll DNS for updates for those nodes.

**Comment by bgrant0607** [comment]
...assuming we use PetSet #24030

**Comment by chrislovecnm** [comment]
@bgrant0607 was talking on the slack dev channel and one of the developers in there was not actually not too concerned. I am thinking that we keep the same API call, but limit the number of seeds saved.

**Comment by chrislovecnm** [comment]
@bgrant0607 @bprashanth we have a comment in the Provider to update some code once the CA services are setup correctly.

Has this been done??

Here is the TODO note: https://github.com/kubernetes/kubernetes/blob/master/examples/cassandra/java/src/io/k8s/cassandra/KubernetesSeedProvider.java#L136

**Comment by bgrant0607** [comment]
@chrislovecnm No idea, unfortunately. Perhaps @brendandburns knows?

**Comment by chrislovecnm** [comment]
Keeping this open until we move the SeedProvider into C* code base.

**Comment by chrislovecnm** [comment]
@bgrant0607 you can assign this to me if you like.

I am adding SnitchProvider design to this document.

- GCE snitch provider needs testing
- EC2 snitch providers need testing (don't have environment yet)
- Kubernetes snitch provider needs design and implementation.

cc: @paralin @mward29 

There are a few different challenges:

The seed provider is not DC aware. You need a seed in every DC, and all nodes need to see the other seeds.

Setting up a snitch can depend on your cloud deployment. With affinity rules or node labels you can pretty much have a balance deployment across multiple zones in AWS or GCE.
We could hack in a SimpleSnitch by accessing node labels from the pod itself.
Rolling restarts are not supported yet. Will be soon.
Probably will open an issue in the kubernetes project and continue the discussion there.

tldr; DCs across multiple K8 instances is not supported, cannot use load balancers. Need to put in an issue about this with the kubernetes project. Seed provider needs some design and some work to support multiple DC's. Multiple racks will probably work, but not tested.

**Comment by paralin** [comment]
So I'm doing this now by joining the networks of two locations and then manually setting the remote-datacenter IP addresses when bootstrapping the cluster. For the frequency I add datacenters (somewhat frequent) I'd like a way to do this automatically. But at the same time not all of the remote datacenters I'm adding are k8s-enabled.

A simple API to query for running k8s instances would be nice. And I will probably write this myself in the meantime.

**Comment by chrislovecnm** [comment]
@paralin we already have a custom SeedProvider ... and I would like to have a custom Snitch.  I have already be told that we can get get both into the C* codebase.   The GoogleCloud Snitch should already work (have not tested).  That could be a good starting place.

Google Snitch https://github.com/apache/cassandra/blob/trunk/src/java/org/apache/cassandra/locator/GoogleCloudSnitch.java

K8s Seed Provider:

https://github.com/kubernetes/kubernetes/blob/master/examples/cassandra/java/src/main/java/io/k8s/cassandra/KubernetesSeedProvider.java

The thing I wish for is Self Discovery via K8s api.  One click automation ... no manual config ... create the yaml and push enter!

Let me know what you think, and please improve via PR!!   Once it is stable, then I want to have Datastax get it into the C* code base.

**Comment by paralin** [comment]
I suppose having this even when you're not running in a k8s pod is fine assuming the k8s api is reachable.

Is the seed code only used at bootstrapping time? So what I'm wondering is, if I lost connectivity to the k8s cluster, could my "remote datacenter" still keep everything running locally?

**Comment by chrislovecnm** [comment]
@paralin seeds are used to bootstrap and when gossip goes ass over elbows.  

use in:
Start-up / Bootstrap / Gossip / OMG the sky is falling and I cannot find my peers via gossip

You probably know this already, but good place to document:

1. Do not make every C* node a seed
2. replacing a seed is a pain (have not worked on how the hell to do that on on k8s ... manual intervention ... testing .. may work now)
3. you need 2 seeds in every datacenter
4. you do not need a seed in every rack

Snitch tells a cassandra node:
1. DC I am in
2. Rack I am in

The keyspace schema defines how this works at data level.

Yo cannot rename DCs.
You can add more DCs

Latency between DC's is EVIL and leads to many repairs.

SeedProvider needs to be DC aware to provide 2 Seeds in every DC

SnitchProvider - k8s needs one that provides discovery via K8S api, and possibly across one of more K8s instances.  Security inside one instance is trivial, multiples can be a pain, but possible.  

Google Snitch example - Every Geolocation is a DC, Every availability zone is a rack.  Not fine grained enough.   We will have use cases to have 2 DCs in the same availability zone.

KubernetesSnitch ideas -
- node labels
- EVN vars
- e2e testing in K8s

KubernetesSeeds Ideas -
- a K8s headless service for each C* DC
- This can be done, super easy with Seed code
- also the seed code needs to use the apache http client, not java http ... C* already depends on Apache HTTP.
- e2e testing in K8s

We can use the Properties file Snitch as long as you launch the Seeds quickly.   Technically we could use dns instead of ips, but that is another dependency ... 

I think utilizing the K8s API for discovery is a must.  Otherwise it is quirky,  complicated and requires manual intervention.  All those things are completely against my beliefs in DevOps ... But I am a pissy, demanding, and don't want to be paged at 3 am.

Note for non-cassandra peeps:

- Cassandra (C*)
- DC in this context is a Cassandra Datacenter

**Comment by chrislovecnm** [comment]
This needs to get into a PR, so that it can be a working document ...

**Comment by paralin** [comment]
I thought Cassandra was/is designed to be used for cross-locational replication? Like at Netflix, they replicate their cassandra data between AWS regions. If latency is such a killer, how does that work?

**Comment by chrislovecnm** [comment]
@paralin can be a pain :)  Just don't use a straw with thousands of nodes

**Comment by chrislovecnm** [comment]
https://github.com/kubernetes/kubernetes/pull/30577 Includes initial support

**Comment by VinceMD** [comment]
@chrislovecnm sorry this is closed but regarding `EC2 snitch providers need testing (don't have environment yet)"`.
I've tested it and it works

In statefulset, remove env var for Rack and DC and instead add this env var:

`          - name: CASSANDRA_ENDPOINT_SNITCH
            value: "Ec2Snitch"`

Result with nodetool status:

`root@cassandra-0:/# nodetool status                                                                                                                                                                   
Datacenter: ap-southeast-2
==========================
Status=Up/Down
|/ State=Normal/Leaving/Joining/Moving
--  Address     Load       Tokens       Owns (effective)  Host ID                               Rack
UN  10.2.37.22  99.61 KiB  32           68.5%             50a6fd85-4443-4e00-96bd-d10c4d719f82  2a
UN  10.2.8.21   99.62 KiB  32           58.2%             1f92206a-8acd-4381-9c9c-38847ecbb202  2b
UN  10.2.131.9  99.62 KiB  32           73.3%             6ea39044-b30e-405b-a7fd-d07b7daf0714  2c`

**Comment by VinceMD** [comment]
With regards to Seed, how to I specify the number of Seeds with current statefulset? Where is the ticket for the development of the SeedProvider in Kubernetes?
